packages/amplify-category-function/src/provider-utils/awscloudformation/utils/environmentVariablesHelper.ts

```diff
@@ -0,0 +1,116 @@
+    const cameledKey = _.camelCase(key);
+    newList.push({
+      cloudFormationParameterName: cameledKey,
+      environmaneVariableName: key,
+    });
+    newReferences[key] = { Ref: cameledKey };
+    newKeyValue[cameledKey] = value;
+  });
+
+  updateStoredList(resourceName, newList);
+  updateStoredReferences(resourceName, newReferences);
+  updateStoredKeyValue(context, resourceName, newKeyValue);
+};
+
+export const deleteEnvironmentVariable = (context: $TSContext, resourceName: string, targetedKey: string): void => {
+  let newList = getStoredList(resourceName);
+  const newReferences = getStoredReferences(resourceName);
+  const newKeyValue = getStoredKeyValue(context, resourceName);
+  const cameledKey = _.camelCase(targetedKey);
+  newList = _.filter(newList, item => {
+    return item.cloudFormationParameterName !== cameledKey && item.environmaneVariableName !== targetedKey;
+  });
+  _.unset(newReferences, targetedKey);
+  _.unset(newKeyValue, cameledKey);
+
+  updateStoredList(resourceName, newList);
+  updateStoredReferences(resourceName, newReferences);
+  updateStoredKeyValue(context, resourceName, newKeyValue);
+};
+
+const getStoredList = (resourceName: string): { cloudFormationParameterName: string; environmaneVariableName: string }[] => {
+  const projectBackendDirPath = pathManager.getBackendDirPath();
+  const resourcePath = path.join(projectBackendDirPath, categoryName, resourceName);
+  const functionParameterFilePath = path.join(resourcePath, functionParametersFileName);
+  const functionParameters = (JSONUtilities.readJson(functionParameterFilePath, { throwIfNotExist: false }) as object) || {};
+  return _.get(functionParameters, 'environmentVariableList', []);
+};
+
+const updateStoredList = (resourceName: string, newList: object): void => {
+  const projectBackendDirPath = pathManager.getBackendDirPath();
+  const resourcePath = path.join(projectBackendDirPath, categoryName, resourceName);
+  const functionParameterFilePath = path.join(resourcePath, functionParametersFileName);
+  const functionParameters = (JSONUtilities.readJson(functionParameterFilePath, { throwIfNotExist: false }) as object) || {};
+  _.set(functionParameters, 'environmentVariableList', newList);
+  JSONUtilities.writeJson(functionParameterFilePath, functionParameters);
+};
+
+const getStoredReferences = (resourceName: string): object => {
+  const projectBackendDirPath = pathManager.getBackendDirPath();
+  const resourcePath = path.join(projectBackendDirPath, categoryName, resourceName);
+  const cfnFileName = `${resourceName}-cloudformation-template.json`;
+  const cfnFilePath = path.join(resourcePath, cfnFileName);
+  const cfnContent = JSONUtilities.readJson(cfnFilePath, { throwIfNotExist: false }) || {};
+  return _.get(cfnContent, ['Resources', 'LambdaFunction', 'Properties', 'Environment', 'Variables'], {});
+};
+
+const updateStoredReferences = (resourceName: string, newReferences: object): void => {
```

Comment:
instead of relying on the caller of these functions to update newReferences properly, it would be better if this function was `setFunctionEnvVarReference = (resourceName: string, envVarName: string, cfnParameter: string)` and this function handled adding the ref and the parameter to the template directly